modules/core/05-port/types/ibc_legacy_module.go

```diff
@@ -228,13 +232,70 @@ func (im *LegacyIBCModule) OnSendPacket(
 // is returned if the packet data is successfully decoded and the receive application
 // logic returns without error.
 // A nil acknowledgement may be returned when using the packet forwarding feature. This signals to core IBC that the acknowledgement will be written asynchronously.
-func (*LegacyIBCModule) OnRecvPacket(
+func (im *LegacyIBCModule) OnRecvPacket(
 	ctx sdk.Context,
 	channelVersion string,
 	packet channeltypes.Packet,
 	relayer sdk.AccAddress,
 ) ibcexported.RecvPacketResult {
-	return ibcexported.RecvPacketResult{}
+	var resultList []ibcexported.RecvPacketResult
+	cbs := im.reversedCallbacks()
+	for _, cb := range cbs {
+		cbVersion := channelVersion
+
+		if wrapper, ok := cb.(VersionWrapper); ok {
+			cbVersion, channelVersion = wrapper.UnwrapVersionSafe(ctx, packet.DestinationPort, packet.DestinationChannel, cbVersion)
+		}
+
+		res := cb.OnRecvPacket(ctx, cbVersion, packet, relayer)
+		resultList = append(resultList, res)
+	}
```

Comment:
Food for thought here. 
I'm thinking that maybe this logic is all that needs to be done within the `LegacyIBCModule`.

For both backwards compat / forwards compat(multi packetdata/ack) we will need to write the temporary results to state if handling an async packet. 

Therefore, I think `LegacyIBCModule` should return a list of results and it should be handled by core. Acknowledgement data can be written to state or glued back together via `WrapAcknowledgement` if it is not async. 
Note, we run into problems trying to do store writes in here anyways, as we do not have access to channelKeeper or ibc store key unless plumbing it through the router (where `LegacyIBCModule` is instantiated).

We would then handle the differing apis in core temporarily while the legacy module type is phased out.

```diff
diff --git a/modules/core/05-port/types/ibc_legacy_module.go b/modules/core/05-port/types/ibc_legacy_module.go
index 6cf2ab0f2..cda632622 100644
--- a/modules/core/05-port/types/ibc_legacy_module.go
+++ b/modules/core/05-port/types/ibc_legacy_module.go
@@ -237,7 +237,7 @@ func (im *LegacyIBCModule) OnRecvPacket(
        channelVersion string,
        packet channeltypes.Packet,
        relayer sdk.AccAddress,
-) ibcexported.RecvPacketResult {
+) []ibcexported.RecvPacketResult {
        var resultList []ibcexported.RecvPacketResult
        cbs := im.reversedCallbacks()
        for _, cb := range cbs {
diff --git a/modules/core/keeper/msg_server.go b/modules/core/keeper/msg_server.go
index 00ab0e398..336704b48 100644
--- a/modules/core/keeper/msg_server.go
+++ b/modules/core/keeper/msg_server.go
@@ -467,6 +467,17 @@ func (k *Keeper) RecvPacket(goCtx context.Context, msg *channeltypes.MsgRecvPack
        //
        // Cache context so that we may discard state changes from callback if the acknowledgement is unsuccessful.
        cacheCtx, writeFn = ctx.CacheContext()
+
+       var results []exported.RecvPacketResult
+       if legacyIBCModule, ok := cbs[0].(porttypes.ClassicIBCModule); ok {
+               results = legacyIBCModule.OnRecvPacket(ctx, channelVersion, msg.Packet, relayer)
+       } else {
+               for _, cb := range cbs {
+                       res := cb.OnRecvPacket(ctx, channelVersion, msg.Packet, relayer)
+                       results = append(results, res)
+               }
+       }
+
```

Reply:
Implemented